// @devexpress/analytics-core/query-builder/utils/_controlsFactory.js
/**
* DevExpress Analytics (query-builder\utils\_controlsFactory.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerControls = void 0;
var controlRegistrator_1 = require("../../diagram/controlRegistrator");
var controlsFactory_1 = require("./controlsFactory");
var metadata_1 = require("../metadata");
var elementViewModel_1 = require("../../core/elements/elementViewModel");
var baseSurface_1 = require("../../core/elements/baseSurface");
var relationModelMeta_1 = require("../elements/relationModelMeta");
var relationSurface_1 = require("../elements/relationSurface");
var relationModel_1 = require("../elements/relationModel");
var joinConditionModelMeta_1 = require("../elements/joinConditionModelMeta");
var joinConditionSurface_1 = require("../elements/joinConditionSurface");
var joinConditionModel_1 = require("../elements/joinConditionModel");
var tableModelMeta_1 = require("../elements/tableModelMeta");
var tableSurface_1 = require("../elements/tableSurface");
var tableModel_1 = require("../elements/tableModel");
var columnModelMeta_1 = require("../elements/columnModelMeta");
var columnSurface_1 = require("../elements/columnSurface");
var columnModel_1 = require("../elements/columnModel");
var queryModelMeta_1 = require("../elements/queryModelMeta");
var querySurface_1 = require("../elements/querySurface");
var queryModel_1 = require("../elements/queryModel");
var _federationQueryModel_1 = require("../elements/_federationQueryModel");
function registerControls() {
    controlRegistrator_1.registerControls();
    controlsFactory_1.controlsFactory.registerControl('Unknown', {
        info: metadata_1.unknownSerializationsInfo,
        type: elementViewModel_1.ElementViewModel,
        nonToolboxItem: true,
        surfaceType: baseSurface_1.SurfaceElementBase,
        isDeleteDeny: true
    });
    controlsFactory_1.controlsFactory.registerControl('Relation', {
        info: relationModelMeta_1.relationSerializationInfo,
        defaultVal: {},
        surfaceType: relationSurface_1.RelationSurface,
        popularProperties: [],
        type: relationModel_1.RelationViewModel,
        elementActionsTypes: [],
        nonToolboxItem: true
    });
    controlsFactory_1.controlsFactory.registerControl('JoinCondition', {
        info: joinConditionModelMeta_1.joinConditionSerializationInfo,
        defaultVal: {},
        surfaceType: joinConditionSurface_1.JoinConditionSurface,
        popularProperties: ['_parentColumnName', '_nestedColumnName', 'joinType'],
        type: joinConditionModel_1.JoinConditionViewModel,
        elementActionsTypes: [],
        nonToolboxItem: true
    });
    controlsFactory_1.controlsFactory.registerControl('Table', {
        info: tableModelMeta_1.tableSerializationInfo,
        defaultVal: {},
        surfaceType: tableSurface_1.TableSurface,
        popularProperties: ['name', 'alias', 'columns'],
        type: tableModel_1.TableViewModel,
        elementActionsTypes: [],
        nonToolboxItem: true
    });
    controlsFactory_1.controlsFactory.registerControl('Column', {
        info: columnModelMeta_1.columnSerializationInfo,
        defaultVal: {},
        surfaceType: columnSurface_1.ColumnSurface,
        popularProperties: ['name', 'alias', 'selected'],
        type: columnModel_1.ColumnViewModel,
        elementActionsTypes: [],
        nonToolboxItem: true,
        isDeleteDeny: true
    });
    controlsFactory_1.controlsFactory.registerControl('Query', {
        info: queryModelMeta_1.querySerializationsInfo,
        surfaceType: querySurface_1.QuerySurface,
        popularProperties: ['name', 'filterString', 'groupFilterString'],
        type: queryModel_1.QueryViewModel,
        elementActionsTypes: [],
        isContainer: true,
        nonToolboxItem: true,
        isDeleteDeny: true
    });
    controlsFactory_1.controlsFactory.registerControl('FTable', {
        info: tableModelMeta_1.tableSerializationInfo,
        defaultVal: {},
        surfaceType: _federationQueryModel_1.FederationTableSurface,
        popularProperties: ['name', 'alias', 'columns'],
        type: _federationQueryModel_1.FederationTableViewModel,
        elementActionsTypes: [],
        nonToolboxItem: true
    });
    controlsFactory_1.controlsFactory.registerControl('FQuery', {
        info: _federationQueryModel_1.federationQuerySerializationsInfo,
        surfaceType: _federationQueryModel_1.FederationQuerySurface,
        popularProperties: ['name'],
        type: _federationQueryModel_1.FederationQueryViewModel,
        elementActionsTypes: [],
        isContainer: true,
        nonToolboxItem: true,
        isDeleteDeny: true
    });
}
exports.registerControls = registerControls;
