// @devexpress/analytics-core/query-builder/wizard/pageId.js
/**
* DevExpress Analytics (query-builder\wizard\pageId.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FullscreenDataSourceWizardSectionId = exports.FullscreenDataSourceWizardPageId = exports.ObjectDataSourceWizardPageId = exports.JsonDataSourceWizardPageId = exports.FederationDataSourceWizardPageId = exports.SqlDataSourceWizardPageId = exports.DataSourceWizardPageId = void 0;
exports.DataSourceWizardPageId = {
    ChooseDataSourceTypePage: 'chooseDataSourceTypePage',
    ConfigureMasterDetailRelationshipsPage: 'configureMasterDetailRelationshipsPage'
};
exports.SqlDataSourceWizardPageId = {
    ChooseConnectionPage: 'chooseSqlConnectionPage',
    ConfigureQueryPage: 'configureSqlQueryPage',
    ConfigureParametersPage: 'configureSqlParametersPage',
    MultiQueryConfigurePage: 'multiSqlQueryConfigurePage',
    MultiQueryConfigureParametersPage: 'multiSqlQueryConfigureParametersPage',
    FederatedQueryConfigurePage: 'federatedQueryConfigurePage'
};
exports.FederationDataSourceWizardPageId = {
    FederatedQueryConfigurePage: 'federatedQueryConfigurePage',
    FederatedMasterDetailRelationshipsPage: 'federatedMasterDetailRelationshipsPage'
};
exports.JsonDataSourceWizardPageId = {
    ChooseJsonSourcePage: 'chooseJsonSourcePage',
    ChooseJsonSchemaPage: 'chooseJsonSchemaPage',
    ChooseConnectionPage: 'chooseJsonConnectionPage',
    SpecifyJsonConnectionPage: 'specifyJsonConnectionPage'
};
exports.ObjectDataSourceWizardPageId = {
    ChooseTypesPage: 'chooseObjectDataSourceTypesPage',
    ChooseDataMembersPage: 'chooseObjectDataSourceDataMembersPage',
    ConfigureParametersPage: 'configureObjectDataSourceParametersPage'
};
exports.FullscreenDataSourceWizardPageId = {
    ChooseDataSourceTypePage: exports.DataSourceWizardPageId.ChooseDataSourceTypePage,
    SpecifySqlDataSourceSettingsPage: 'specifySqlDataSourceSettingsPage',
    SpecifyJsonDataSourceSettingsPage: 'specifyJsonDataSourceSettingsPage',
    SpecifyObjectDataSourceSettingsPage: 'specifyObjectDataSourceSettingsPage',
    SpecifyFederationDataSourceSettingsPage: 'specifyFederationDataSourceSettingsPage',
};
exports.FullscreenDataSourceWizardSectionId = {
    SpecifyJsonConnectionPage: exports.JsonDataSourceWizardPageId.SpecifyJsonConnectionPage,
    ChooseJsonSchemaPage: exports.JsonDataSourceWizardPageId.ChooseJsonSchemaPage,
    ChooseJsonSourcePage: exports.JsonDataSourceWizardPageId.ChooseJsonSourcePage,
    ChooseSqlConnectionPage: exports.SqlDataSourceWizardPageId.ChooseConnectionPage,
    ConfigureFederatedQueriesPage: exports.FederationDataSourceWizardPageId.FederatedQueryConfigurePage,
    ConfigureFederatedMasterDetailRelationshipsPage: exports.FederationDataSourceWizardPageId.FederatedMasterDetailRelationshipsPage,
    ConfigureQueryPage: exports.SqlDataSourceWizardPageId.MultiQueryConfigurePage,
    ConfigureQueryParametersPage: exports.SqlDataSourceWizardPageId.MultiQueryConfigureParametersPage,
    ConfigureMasterDetailRelationshipsPage: exports.DataSourceWizardPageId.ConfigureMasterDetailRelationshipsPage
};
