// @devexpress/analytics-core/query-builder/wizard/pageFactory.js
/**
* DevExpress Analytics (query-builder\wizard\pageFactory.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PageFactory = void 0;
var PageFactory = (function () {
    function PageFactory() {
        this.metadata = {};
    }
    PageFactory.prototype.registerMetadata = function (pageId, metadata) {
        if (!metadata.canFinish)
            metadata.canFinish = function (page) { return page.canFinish(); };
        if (!metadata.canNext)
            metadata.canNext = function (page) { return page.canNext(); };
        this.metadata[pageId] = metadata;
    };
    PageFactory.prototype.getMetadata = function (pageId) {
        return this.metadata[pageId];
    };
    PageFactory.prototype.unregisterMetadata = function (pageId) {
        delete this.metadata[pageId];
    };
    PageFactory.prototype.reset = function () {
        this.metadata = {};
    };
    return PageFactory;
}());
exports.PageFactory = PageFactory;
