// @devexpress/analytics-core/widgets/criteria/operators/group.js
/**
* DevExpress Analytics (widgets\criteria\operators\group.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.GroupOperator = void 0;
var criteriaOperator_1 = require("./criteriaOperator");
var criteriaOperatorPreprocessor_1 = require("../utils/criteriaOperatorPreprocessor");
var group_1 = require("./options/group");
var binary_1 = require("./options/binary");
var GroupOperator = (function (_super) {
    __extends(GroupOperator, _super);
    function GroupOperator(operation, operands) {
        var _this = _super.call(this) || this;
        _this.create = function (isGroup, property, specifics) {
            var operator = criteriaOperatorPreprocessor_1.criteriaCreator.process('binary', {
                left: property,
                right: criteriaOperatorPreprocessor_1.criteriaCreator.process('value', { value: '' }),
                operatorType: binary_1.BinaryOperatorType.Equal
            });
            if (isGroup) {
                operator = criteriaOperatorPreprocessor_1.criteriaCreator.process('group', { operation: group_1.GroupOperatorType.And, operands: [] });
            }
            else if (specifics && specifics === 'list') {
                operator = criteriaOperatorPreprocessor_1.criteriaCreator.process('aggregate', {
                    aggregateType: 'Exisits',
                    property: property,
                    aggregatedExpression: null,
                    condition: criteriaOperatorPreprocessor_1.criteriaCreator.process('group', {
                        operands: [],
                        operation: group_1.GroupOperatorType.And
                    })
                });
            }
            _this.operands.push(operator);
            return _this.operands[_this.operands.indexOf(operator)];
        };
        _this.change = function (operationType, item, incorrectSpecificsForAggregate) {
            if (incorrectSpecificsForAggregate === void 0) { incorrectSpecificsForAggregate = false; }
            var position = _this.operands.indexOf(item);
            if (position !== -1) {
                var operator = criteriaOperatorPreprocessor_1.criteriaCreator.changeByType(operationType);
                if (operationType.type !== operator.enumType && !Array.isArray(operator.leftPart)) {
                    operator.leftPart.assignFrom(item, incorrectSpecificsForAggregate, !operationType.emptyRightPart);
                }
                else {
                    operator.assignFrom(item, incorrectSpecificsForAggregate, !operationType.emptyRightPart);
                }
                _this.operands[position] = operator;
            }
            else {
                throw Error('dont have this element in operands collection');
            }
            item = null;
            return _this.operands[position];
        };
        _this.remove = function (operator) {
            _this.operands.splice(_this.operands.indexOf(operator), 1);
        };
        _this.assignLeftPart = function (operator) {
            _this.operands = operator.operands;
        };
        _this.operands = [];
        _this.type = 'group';
        _this.operatorType = operation;
        operands = operands || [criteriaOperatorPreprocessor_1.criteriaCreator.process('default'), criteriaOperatorPreprocessor_1.criteriaCreator.process('default')];
        operands.forEach(function (operand) { return _this.operands.push(operand); });
        return _this;
    }
    GroupOperator.combine = function (operation, operands) {
        var combinedOperands = [];
        (operands || []).forEach(function (operand) {
            if (operand.type === 'group' && operand.operatorType === operation) {
                combinedOperands.push.apply(combinedOperands, operand.operands);
            }
            else {
                combinedOperands.push(operand);
            }
        });
        if (combinedOperands.length === 1) {
            return combinedOperands[0];
        }
        return criteriaOperatorPreprocessor_1.criteriaCreator.process('group', { operands: combinedOperands, operation: operation });
    };
    GroupOperator.prototype.children = function () {
        return this.operands;
    };
    Object.defineProperty(GroupOperator.prototype, "displayType", {
        get: function () {
            return group_1.GroupOperatorType[this.operatorType];
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(GroupOperator.prototype, "enumType", {
        get: function () {
            return group_1.GroupOperatorType;
        },
        enumerable: false,
        configurable: true
    });
    GroupOperator.prototype.accept = function (visitor) {
        return visitor.visitGroupOperator
            ? visitor.visitGroupOperator(this)
            : new GroupOperator(this.operatorType, this.operands && this.operands.map(function (op) { return op.accept(visitor); }));
    };
    return GroupOperator;
}(criteriaOperator_1.CriteriaOperator));
exports.GroupOperator = GroupOperator;
criteriaOperatorPreprocessor_1.criteriaCreator.register('group', function (options) {
    return new GroupOperator(options.operation, options.operands);
}, function (operatorType) { return criteriaOperatorPreprocessor_1.criteriaCreator.process('group', {
    operation: operatorType.value,
    operands: []
}); });
