// @devexpress/analytics-core/query-builder/wizard/internal/federationDataSource/_federationDataSourceItemsExtender.js
/**
* DevExpress Analytics (query-builder\wizard\internal\federationDataSource\_federationDataSourceItemsExtender.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FederationDataSourceItemsExtender = void 0;
var FederationDataSourceItemsExtender = (function () {
    function FederationDataSourceItemsExtender(_rootItems) {
        this._rootItems = _rootItems;
    }
    FederationDataSourceItemsExtender.prototype.afterItemsFilled = function (request, items) {
        items.forEach(function (x) {
            if (x.isListType === undefined) {
                x.isListType = x.specifics === 'List';
            }
        });
    };
    FederationDataSourceItemsExtender.prototype.beforeItemsFilled = function (request, items) {
        if (!request.fullPath) {
            items.push.apply(items, this._rootItems().map(function (item) {
                return {
                    name: item.id || item.ref,
                    displayName: item.name,
                    isList: true,
                    specifics: item.specifics || 'ListSource',
                    isSupportQueries: item.isSupportQueries,
                    isListType: item.isListType
                };
            }));
            return true;
        }
        return false;
    };
    return FederationDataSourceItemsExtender;
}());
exports.FederationDataSourceItemsExtender = FederationDataSourceItemsExtender;
