// @devexpress/analytics-core/core/dragDrop/_wizardDragDropHandler.js
/**
* DevExpress Analytics (core\dragDrop\_wizardDragDropHandler.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WizardDragDropHandler = void 0;
var ko = require("knockout");
var $ = require("jquery");
var _dragDropHandler_1 = require("./_dragDropHandler");
var rectangle_1 = require("../elements/rectangle");
var templateUtils_1 = require("../../property-grid/widgets/templateUtils");
var WizardDragDropHandler = (function (_super) {
    __extends(WizardDragDropHandler, _super);
    function WizardDragDropHandler(options) {
        var _this = _super.call(this, null, null, ko.observable(null), null, options.dragHelperContent) || this;
        _this.parent = function () { return $(options.parent); };
        _this.containment = options.containment;
        _this._target = options.target;
        _this._addHandler = options.addHandler;
        return _this;
    }
    WizardDragDropHandler.prototype.helper = function (draggable, event) {
        this._dropTarget = null;
        var item = draggable;
        var target = $(event.target).closest('.ui-draggable');
        var rect = new rectangle_1.Rectangle(event.offsetX + 6, event.offsetY + 6, target.width(), target.height());
        rect.className = 'dxrd-image-ghost-report';
        this.dragHelperContent.reset();
        this.dragHelperContent.setContent(rect, {
            template: 'dxrd-drag-helper-source-reorder-treelist',
            data: {
                imageClassName: item.imageClassName,
                imageTemplateName: item.imageTemplateName,
                text: item.text
            }
        });
        var templateHtml = templateUtils_1.getTemplate(this.dragHelperContent.template);
        var $container = $(templateHtml).css({ 'display': 'block' });
        $container.prependTo(this.parent());
        ko.applyBindingsToDescendants(this.dragHelperContent, $container[0]);
        return $container;
    };
    WizardDragDropHandler.prototype.doStopDrag = function (ui, draggable) {
        this.dragHelperContent.reset();
        if (this._dropTarget) {
            this._addHandler(this._dropTarget, draggable, this._getAbsoluteSurfacePosition(ui));
        }
    };
    WizardDragDropHandler.prototype.drag = function (event, ui) {
        var $target = $(this.getTarget(event));
        var page = $target.closest(this._target);
        this._dropTarget = null;
        if (page.length && (!this._target || $target.closest(this._target)[0])) {
            this._dropTarget = ko.dataFor(page.get(0));
        }
    };
    return WizardDragDropHandler;
}(_dragDropHandler_1.DragDropHandler));
exports.WizardDragDropHandler = WizardDragDropHandler;
