// @devexpress/analytics-core/query-builder/wizard/fullscreen/pages/specifyFederationDataSourceSettingsPage.js
/**
* DevExpress Analytics (query-builder\wizard\fullscreen\pages\specifyFederationDataSourceSettingsPage.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports._registerSpecifyFederationDataSourceSettingsPage = exports.SpecifyFederationDataSourceSettingsPage = void 0;
var fullscreenWizardPage_1 = require("./fullscreenWizardPage");
var pageId_1 = require("../../pageId");
var dataSourceWizardState_1 = require("../../dataSourceWizardState");
var localization_utils_1 = require("../../../../property-grid/localization/localization_utils");
var _utils_1 = require("../../../../serializer/_utils");
var federatedQueryConfigurePage_1 = require("../../pages/federationDataSourceWizard/federatedQueryConfigurePage");
var federatedMasterDetailRelationshipsPage_1 = require("../../pages/federationDataSourceWizard/federatedMasterDetailRelationshipsPage");
var _requestwrapper_1 = require("../../../utils/_requestwrapper");
var _utils_2 = require("../../internal/_utils");
var SpecifyFederationDataSourceSettingsPage = (function (_super) {
    __extends(SpecifyFederationDataSourceSettingsPage, _super);
    function SpecifyFederationDataSourceSettingsPage(_dataSourceWizardOptions) {
        var _this = _super.call(this) || this;
        _this._dataSourceWizardOptions = _dataSourceWizardOptions;
        return _this;
    }
    SpecifyFederationDataSourceSettingsPage.prototype.getNextSectionId = function (sectionId) {
        if (!sectionId)
            return pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedQueriesPage;
        else if (sectionId === pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedQueriesPage) {
            var federationDataSourceWizard = dataSourceWizardState_1._restoreFederationDataSourceFromState(this._stateManager.getCurrentState().federationDataSourceWizard, this._dataSourceWizardOptions.dataSources);
            if (federationDataSourceWizard.queries().length > 1) {
                return pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedMasterDetailRelationshipsPage;
            }
        }
    };
    SpecifyFederationDataSourceSettingsPage.prototype._showPageDescription = function () {
        return true;
    };
    SpecifyFederationDataSourceSettingsPage.prototype.registerSections = function () {
        federatedQueryConfigurePage_1._registerFederatedQueryConfigurePage(this._factory, this._dataSourceWizardOptions);
        federatedMasterDetailRelationshipsPage_1._registerFederatedMasterDetailRelationshipsPage(this._factory, _requestwrapper_1.wrapGetFederationdResultSchema(this._dataSourceWizardOptions.callbacks.federationDataSourceResultSchema), this._dataSourceWizardOptions);
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedQueriesPage);
        meta.description = localization_utils_1.getLocalization('Create a federated query.', 'DataAccessUIStringId.WizardPageConfigureFederatedQueryPage');
        var meta = this._factory.getMetadata(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedMasterDetailRelationshipsPage);
        meta.description = localization_utils_1.getLocalization('Configure master-detail relationships.', 'AnalyticsCoreStringId.SqlDSWizard_PageConfigureMasterDetailRelations');
        meta['disabledText'] = localization_utils_1.getLocalization('To create a master-detail relationship, select two or more queries.', 'AnalyticsCoreStringId.Wizard_MasterDetailRelationship_Placeholder');
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedQueriesPage, _utils_2.WizardSectionPosition.Top);
        this._setSectionPosition(pageId_1.FullscreenDataSourceWizardSectionId.ConfigureFederatedMasterDetailRelationshipsPage, _utils_2.WizardSectionPosition.Bottom);
    };
    return SpecifyFederationDataSourceSettingsPage;
}(fullscreenWizardPage_1.FullscreenWizardPage));
exports.SpecifyFederationDataSourceSettingsPage = SpecifyFederationDataSourceSettingsPage;
function _registerSpecifyFederationDataSourceSettingsPage(factory, dataSourceWizardOptions) {
    factory.registerMetadata(pageId_1.FullscreenDataSourceWizardPageId.SpecifyFederationDataSourceSettingsPage, {
        setState: function (data, state) {
            state.federationDataSourceWizard.federationDataSourceJSON = data.federationDataSourceWizard.federationDataSourceJSON;
            state.federationDataSourceWizard.federatedQueries = data.federationDataSourceWizard.federatedQueries;
            state.federationDataSourceWizard.relations = data.federationDataSourceWizard.relations;
            state.federationDataSourceWizard.name = data.federationDataSourceWizard.name;
        },
        getState: function (state) { return state; },
        resetState: function (state, defaulState) {
            state.federationDataSourceWizard = _utils_1.extend(true, {}, defaulState);
        },
        create: function () {
            return new SpecifyFederationDataSourceSettingsPage(dataSourceWizardOptions);
        },
        navigationPanelText: localization_utils_1.getLocalization('Specify Data Source Settings', 'AnalyticsCoreStringId.Wizard_SpecifyDataSourceSettingsPage'),
        template: 'dx-wizard-fullscreen-page'
    });
}
exports._registerSpecifyFederationDataSourceSettingsPage = _registerSpecifyFederationDataSourceSettingsPage;
