// devexpress-reporting/designer/internal/parameterLayout/_parameterLayoutDragDropHandler.js
/**
* DevExpress HTML/JS Reporting (designer\internal\parameterLayout\_parameterLayoutDragDropHandler.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParameterLayoutDragDropHandler = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var $ = require("jquery");
var ko = require("knockout");
var layoutItems_1 = require("../../dataObjects/parameters/layoutItems");
var _objectExplorerDragDropHandler_1 = require("../dragdrop/_objectExplorerDragDropHandler");
var _objectExplorerDragDropHelper_1 = require("../dragdrop/_objectExplorerDragDropHelper");
var ParameterLayoutDragDropHelper = (function (_super) {
    __extends(ParameterLayoutDragDropHelper, _super);
    function ParameterLayoutDragDropHelper(_selectedItem, dragHelperContent) {
        var _this = _super.call(this, dragHelperContent) || this;
        _this._selectedItem = _selectedItem;
        _this._dropBefore = false;
        _this._dropInside = false;
        return _this;
    }
    ParameterLayoutDragDropHelper.prototype._getDroppableClassName = function (isInTopOrderArea, isInBottomOrderArea) {
        if (!this.canDrop()) {
            return this.droppableClassName;
        }
        this._dropInside = false;
        var className = this.droppableClassName + ' ' + this.approveClassName;
        if (isInBottomOrderArea) {
            this._dropBefore = true;
            return className + ' ' + this.classDropAfter;
        }
        else if (isInTopOrderArea) {
            this._dropBefore = false;
            return className + ' ' + this.classDropBefore;
        }
        else if (this._targetModel instanceof layoutItems_1.GroupLayoutItem) {
            this._dropInside = true;
            return className;
        }
        return '';
    };
    ParameterLayoutDragDropHelper.prototype.getSiblings = function () {
        return this._draggableModel.parentModel().parameterPanelLayoutItems;
    };
    ParameterLayoutDragDropHelper.prototype.getNewParentModel = function () {
        return this._dropInside ? this._targetModel : this._targetModel.parentModel();
    };
    ParameterLayoutDragDropHelper.prototype.getTargetSiblings = function () {
        return this.getNewParentModel().parameterPanelLayoutItems;
    };
    ParameterLayoutDragDropHelper.prototype.reorderSiblings = function (isDragToBottom) {
        if (isDragToBottom === void 0) { isDragToBottom = this.isDragToBottom(); }
        var siblings = this.getSiblings();
        var targetSiblings = this.getTargetSiblings();
        if (siblings && targetSiblings) {
            var _siblings = siblings.peek();
            var _targetSiblings = targetSiblings.peek();
            _siblings.splice(siblings.indexOf(this._draggableModel), 1);
            var targetIndex = _targetSiblings.indexOf(this._targetModel);
            targetIndex === -1 && targetIndex++;
            _targetSiblings.splice(targetIndex + (isDragToBottom ? 1 : 0), 0, this._draggableModel);
            this._draggableModel.parentModel(this.getNewParentModel());
            siblings.valueHasMutated();
            siblings !== targetSiblings && targetSiblings.valueHasMutated();
            this._selectedItem(this._draggableModel);
        }
    };
    ParameterLayoutDragDropHelper.prototype.canDrop = function () {
        var currentParent = this._targetModel;
        while (currentParent instanceof layoutItems_1.ParameterPanelLayoutItem) {
            if (this._draggableModel === currentParent)
                return false;
            currentParent = currentParent.parentModel();
        }
        return true;
    };
    ParameterLayoutDragDropHelper.prototype.stop = function () {
        _super.prototype.stop.call(this);
        if (!this._target || !this._targetModel || !this.canDrop()) {
            return;
        }
        this.reorderSiblings(this._dropBefore);
    };
    return ParameterLayoutDragDropHelper;
}(_objectExplorerDragDropHelper_1.ObjectExplorerDragDropHelper));
var ParameterLayoutDragDropHandler = (function (_super) {
    __extends(ParameterLayoutDragDropHandler, _super);
    function ParameterLayoutDragDropHandler(selectedItem) {
        var _this = _super.call(this, ko.observable(false), null, null, ko.observable(null), new analytics_internal_1.DragHelperContent(null)) || this;
        _this.containment = '.dxrd-parameters-content-list';
        _this.parent = function () { return $('.dx-designer-viewport .dxrd-parameters-edit-dialog > .dx-popup-normal'); };
        _this.reportControlsDragDropHelper = new ParameterLayoutDragDropHelper(selectedItem, _this.dragHelperContent);
        _this.helper = function (draggable, event) {
            _this.reportControlsDragDropHelper.helper(draggable, event);
            var templateHtml = analytics_widgets_1.getTemplate(_this.dragHelperContent.template);
            var $container = $(templateHtml).css({ 'display': 'block' });
            $container.prependTo(_this.parent());
            ko.applyBindingsToDescendants(_this.dragHelperContent, $container[0]);
            return $container;
        };
        return _this;
    }
    ParameterLayoutDragDropHandler.prototype.startDrag = function (draggable) {
        this.reportControlsDragDropHelper.start(draggable);
        _super.prototype.startDrag.call(this, draggable);
    };
    ParameterLayoutDragDropHandler.prototype.doStopDrag = function (ui, draggable, event) {
        this.reportControlsDragDropHelper.started && this.reportControlsDragDropHelper.clearDroppableClasses();
        this.dragHelperContent.reset();
        this.reportControlsDragDropHelper.stop();
    };
    return ParameterLayoutDragDropHandler;
}(_objectExplorerDragDropHandler_1.ObjectExplorerDragDropHandler));
exports.ParameterLayoutDragDropHandler = ParameterLayoutDragDropHandler;
