// devexpress-reporting/designer/tools/dialogs/parametersDialogs.js
/**
* DevExpress HTML/JS Reporting (designer\tools\dialogs\parametersDialogs.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AddParameterDialog = exports.ParametersDialogBase = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var ko = require("knockout");
var parameter_1 = require("../../dataObjects/metadata/parameters/parameter");
var layoutItems_1 = require("../../dataObjects/parameters/layoutItems");
var parameter_2 = require("../../dataObjects/parameters/parameter");
var parameterExpressionAddon_1 = require("../../dataObjects/parameters/parameterExpressionAddon");
var SettingsAreaModel = (function () {
    function SettingsAreaModel(_parameter) {
        this._parameter = _parameter;
        this.valueSourceSettingsType = this._parameter.valueSourceSettingsType;
        this.valueSourceSettings = this._parameter.valueSourceSettings;
    }
    SettingsAreaModel.prototype.getInfo = function () {
        return this._parameter.getInfo().reduce(function (currentInfo, x) {
            if (x.propertyName === 'valueSourceSettingsType') {
                var info = analytics_internal_1.extend(true, {}, x);
                info.editor = analytics_widgets_1.editorTemplates.getEditor('combobox');
                currentInfo.push(info);
            }
            else if (x.propertyName === parameter_1.valueSourceSettingsSerializationInfo.propertyName) {
                var info = analytics_internal_1.extend(true, {}, x);
                info.editor = analytics_widgets_1.editorTemplates.getEditor('inplaceObjectEditor');
                currentInfo.push(info);
            }
            return currentInfo;
        }, []);
    };
    SettingsAreaModel.prototype.isPropertyVisible = function (propertyName) {
        if (propertyName === 'valueSourceSettings')
            return this.valueSourceSettingsType() !== 'None';
        return true;
    };
    return SettingsAreaModel;
}());
var ParametersDialogBase = (function (_super) {
    __extends(ParametersDialogBase, _super);
    function ParametersDialogBase(_currentReport) {
        var _this = _super.call(this) || this;
        _this._currentReport = _currentReport;
        _this._undoEngine = null;
        _this._isSubmited = false;
        _this.buttons = [
            _this._createButton(analytics_utils_1.getLocalization('OK', analytics_internal_1.StringId.DataAccessBtnOK), function () { return _this.submit(); }, 'default'),
            _this._createButton(analytics_utils_1.getLocalization('Cancel', analytics_internal_1.StringId.DataAccessBtnCancel), function () { return _this.visible(false); })
        ];
        _this._editableObject = ko.observable(null);
        _this._selectedItem = ko.observable(null);
        _this._selectedParameter = ko.observable(null);
        _this._selectedParameterSettings = ko.observable(null);
        _this.visible = ko.observable(false);
        _this.container = function (element) { return analytics_internal_1.getParentContainer(element); };
        _this._disposables.push(_this._propertiesGrid = new analytics_widgets_1.ObjectProperties(_this._editableObject), _this._settingsGrid = new analytics_widgets_1.ObjectProperties(_this._selectedParameterSettings), _this.visible.subscribe(function (newVal) { return !newVal && _this.close(); }), _this._selectedItem.subscribe(function (item) {
            var parameter = item && _this._getParameterFromLayoutItem(item);
            if (parameter) {
                _this._editableObject(parameter);
                _this._selectedParameterSettings(new SettingsAreaModel(parameter));
            }
            else {
                _this._editableObject(item);
                _this._selectedParameterSettings(null);
            }
            _this._selectedParameter(parameter);
        }));
        _this._propertiesGrid.createEditorAddOn = function (editor) {
            if (_this._selectedParameter() && _this._selectedParameter().propertyExpressionMapper.getExpressionProperty(editor.name) &&
                editor.name.toLowerCase().indexOf('value') === -1) {
                var addon = new parameterExpressionAddon_1.ParameterExpressionAddOn(editor, _this._selectedParameter);
                return {
                    data: addon,
                    templateName: 'dx-wizard-menu-box-editorswitch'
                };
            }
        };
        return _this;
    }
    ParametersDialogBase.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this.selectItem(null);
        this.removeProperties();
    };
    ParametersDialogBase.prototype.onSubmit = function () { };
    Object.defineProperty(ParametersDialogBase.prototype, "undoEngine", {
        get: function () {
            if (!this._undoEngine) {
                this._undoEngine = analytics_utils_1.UndoEngine.tryGetUndoEngine(this._currentReport);
            }
            return this._undoEngine;
        },
        enumerable: false,
        configurable: true
    });
    ParametersDialogBase.prototype._getParameterFromLayoutItem = function (layoutItem) {
        return layoutItem instanceof layoutItems_1.ParameterLayoutItem ? layoutItem.parameter() : null;
    };
    ParametersDialogBase.prototype._createParameter = function (parameters) {
        if (parameters === void 0) { parameters = this._currentReport.parameters(); }
        var newName = analytics_internal_1.getUniqueNameForNamedObjectsArray(parameters, 'parameter'), newParameter = new parameter_2.Parameter({ '@Name': newName, '@Description': 'P' + newName.slice(1) }, this._currentReport);
        newParameter._isEditing(true);
        newParameter._showLayoutProperties(true);
        return new layoutItems_1.ParameterLayoutItem({}, this._currentReport, null, newParameter);
    };
    ParametersDialogBase.prototype._createButton = function (text, action, type) {
        if (type === void 0) { type = 'normal'; }
        return {
            toolbar: 'bottom', location: 'after', widget: 'dxButton', options: {
                text: text, type: type, stylingMode: 'contained', onClick: action
            }
        };
    };
    ParametersDialogBase.prototype.selectItem = function (layoutItem) {
        var previousParameter = this._selectedParameter();
        var parameter = this._getParameterFromLayoutItem(layoutItem);
        if (parameter && previousParameter === parameter)
            return;
        this._selectedItem(layoutItem);
    };
    ParametersDialogBase.prototype.show = function (parameter) {
        this.undoEngine.start();
        this._isSubmited = false;
        this._currentReport.parameterHelper.startEditing();
        this._onStart(this._currentReport.parameterHelper.getParameterLayoutItem(parameter));
        this.visible(true);
    };
    ParametersDialogBase.prototype._onStart = function (layoutItem) {
        this.selectItem(layoutItem);
        this._selectedItem.valueHasMutated();
    };
    ParametersDialogBase.prototype.close = function () {
        var cancel = !this._isSubmited && this.undoEngine._hasSessionChanges();
        this._currentReport.parameterHelper.endEditing();
        this.undoEngine.end();
        cancel && this.undoEngine.undo(true);
        this.selectItem(null);
    };
    ParametersDialogBase.prototype.submit = function () {
        this._isSubmited = true;
        this.onSubmit();
        this.visible(false);
    };
    return ParametersDialogBase;
}(analytics_utils_1.Disposable));
exports.ParametersDialogBase = ParametersDialogBase;
var AddParameterDialog = (function (_super) {
    __extends(AddParameterDialog, _super);
    function AddParameterDialog() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.popupCss = 'dxrd-parameters-add-dialog';
        _this.title = analytics_utils_1.getLocalization('Add Parameter', 'ASPxReportsStringId.ReportDesigner_ParametersDialog_AddParameter');
        _this.width = 900;
        _this.height = 632;
        _this.contentTemplate = 'dxrd-parameter-edit';
        return _this;
    }
    AddParameterDialog.prototype.onSubmit = function () {
        this._currentReport.parameterPanelLayoutItems.push(this._selectedItem());
        this._currentReport.parameters.push(this._selectedParameter());
    };
    AddParameterDialog.prototype._onStart = function (parameter) {
        if (parameter === void 0) { parameter = this._createParameter(); }
        _super.prototype._onStart.call(this, parameter);
    };
    return AddParameterDialog;
}(ParametersDialogBase));
exports.AddParameterDialog = AddParameterDialog;
