// @devexpress/analytics-core/query-builder/dataSource/federation/_dbSchemaFederationDataSourceProvider.js
/**
* DevExpress Analytics (query-builder\dataSource\federation\_dbSchemaFederationDataSourceProvider.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.DBSchemaFederationDataSourceProvider = void 0;
var $ = require("jquery");
var utils_1 = require("../../../serializer/utils");
var pathRequest_1 = require("../../../widgets/common/pathRequest");
var dbTable_1 = require("../dbTable");
var DBSchemaFederationDataSourceProvider = (function (_super) {
    __extends(DBSchemaFederationDataSourceProvider, _super);
    function DBSchemaFederationDataSourceProvider(itemsProvider) {
        var _this = _super.call(this) || this;
        _this._rootItems = [];
        _this.getItems = itemsProvider.getItems;
        return _this;
    }
    DBSchemaFederationDataSourceProvider.prototype.getDbSchema = function () {
        var deferred = $.Deferred();
        if (this._rootItems.length)
            deferred.resolve(this._rootItems);
        else {
            this.getItems(new pathRequest_1.PathRequest('')).done(function (rootItems) {
                deferred.resolve(rootItems);
            });
        }
        return deferred.promise();
    };
    DBSchemaFederationDataSourceProvider.prototype.getDbTable = function (tableName, path) {
        var $deferred = $.Deferred();
        this.getItems(new pathRequest_1.PathRequest(path)).done(function (items) {
            var columns = items.map(function (x) {
                return {
                    Name: x.displayName || x.name
                };
            });
            $deferred.resolve(new dbTable_1.DBTable({ Name: tableName, columns: columns }));
        });
        return $deferred.promise();
    };
    DBSchemaFederationDataSourceProvider.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this.getItems = null;
    };
    return DBSchemaFederationDataSourceProvider;
}(utils_1.Disposable));
exports.DBSchemaFederationDataSourceProvider = DBSchemaFederationDataSourceProvider;
