// @devexpress/analytics-core/query-builder/dataSource/federation/federatedQueryExpression.js
/**
* DevExpress Analytics (query-builder\dataSource\federation\federatedQueryExpression.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FederatedQueryExpression = exports.FederatedQueryExpressionType = exports.expressionSerializationInfo = void 0;
var serializer_1 = require("../../../serializer/serializer");
var _utils_1 = require("../../../serializer/_utils");
var metadata_1 = require("../../metadata");
exports.expressionSerializationInfo = [
    { propertyName: 'table', modelName: '@NodeAlias' },
    metadata_1.name,
    metadata_1.alias,
    { propertyName: 'expressionType', modelName: '@ExpressionType' },
    { propertyName: 'columnExpression', modelName: '@ColumnExpression' }
];
var FederatedQueryExpressionType;
(function (FederatedQueryExpressionType) {
    FederatedQueryExpressionType[FederatedQueryExpressionType["SelectColumnExpression"] = 0] = "SelectColumnExpression";
    FederatedQueryExpressionType[FederatedQueryExpressionType["SelectExpression"] = 1] = "SelectExpression";
    FederatedQueryExpressionType[FederatedQueryExpressionType["SelectAllColumnsExpression"] = 2] = "SelectAllColumnsExpression";
    FederatedQueryExpressionType[FederatedQueryExpressionType["SelectAllNodeColumnsExpression"] = 3] = "SelectAllNodeColumnsExpression";
    FederatedQueryExpressionType[FederatedQueryExpressionType["SelectRowCountExpression"] = 4] = "SelectRowCountExpression";
})(FederatedQueryExpressionType = exports.FederatedQueryExpressionType || (exports.FederatedQueryExpressionType = {}));
var FederatedQueryExpression = (function () {
    function FederatedQueryExpression(model, serializer) {
        serializer = serializer || new serializer_1.ModelSerializer();
        serializer.deserialize(this, _utils_1.extend(model, { '@ItemType': 'Expression' }));
    }
    FederatedQueryExpression.prototype.getInfo = function () {
        return exports.expressionSerializationInfo;
    };
    return FederatedQueryExpression;
}());
exports.FederatedQueryExpression = FederatedQueryExpression;
