// devexpress-reporting/designer/internal/dragdrop/_reportExplorerDragDropHandler.js
/**
* DevExpress HTML/JS Reporting (designer\internal\dragdrop\_reportExplorerDragDropHandler.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReportExplorerDragDropHandler = void 0;
var analytics_elements_1 = require("@devexpress/analytics-core/analytics-elements");
var formattingrules_1 = require("../../controls/properties/formattingrules");
var xrReport_1 = require("../../controls/xrReport");
var _objectExplorerDragDropHandler_1 = require("./_objectExplorerDragDropHandler");
var _reportControlsDragDropHelper_1 = require("./_reportControlsDragDropHelper");
var _utils_1 = require("./_utils");
var ReportExplorerDragDropHandler = (function (_super) {
    __extends(ReportExplorerDragDropHandler, _super);
    function ReportExplorerDragDropHandler(canAddItems, surface, selection, undoEngine, dragHelperContent) {
        var _this = _super.call(this, canAddItems, surface, selection, undoEngine, dragHelperContent) || this;
        _this._isStyle = function (item) { return item.data && item.data.specifics === 'stylemodel'; };
        _this._isFormatingRule = function (item) { return item.data && item.data.specifics === 'formattingrule'; };
        _this._isReportControl = function (item) { return !_this._isStyle(item) && !_this._isFormatingRule(item); };
        _this.cursor = 'arrow';
        _this.alwaysAlt = true;
        _this.containment = '.dxrd-designer';
        _this['cursorAt'] = {
            top: 0,
            left: 0
        };
        _this.reportControlsDragDropHelper = new _reportControlsDragDropHelper_1.ReportControlsDragDropHelper(_this.dragHelperContent, _this.undoEngine && _this.undoEngine());
        _this.helper = function (draggable, event) {
            var item = draggable;
            _utils_1.selectTreeListItem(item, event);
            if (_this._isReportControl(item)) {
                _this.reportControlsDragDropHelper.helper(draggable, event);
            }
            else {
                var rect = new analytics_elements_1.Rectangle(12, 12, 12, 12);
                var templateId = '';
                if (_this._isStyle(item)) {
                    rect.className = 'dxrd-image-ghost-stylemodel';
                    templateId = 'dxrd-svg-reportexplorer-style';
                }
                else if (_this._isFormatingRule(item)) {
                    rect.className = 'dxrd-image-ghost-formattingrule';
                    templateId = 'dxrd-svg-reportexplorer-formatting_rule';
                }
                _this.dragHelperContent.reset();
                dragHelperContent.setContent(rect, (templateId ? { template: templateId } : null));
                _this._size.width(12);
                _this._size.height(12);
            }
        };
        return _this;
    }
    ReportExplorerDragDropHandler.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        this.reportControlsDragDropHelper.dispose();
    };
    ReportExplorerDragDropHandler.prototype.startDrag = function (draggable) {
        if (this._isReportControl(draggable)) {
            this.reportControlsDragDropHelper.start(draggable);
        }
        _super.prototype.startDrag.call(this, draggable);
    };
    ReportExplorerDragDropHandler.prototype.doStopDrag = function (ui, draggable, event) {
        this.reportControlsDragDropHelper.started && this.reportControlsDragDropHelper.clearDroppableClasses();
        this.dragHelperContent.reset();
        if (this.dragHelperContent.isLocked())
            return;
        if (this.reportControlsDragDropHelper.started) {
            this.selection.initialize(this.reportControlsDragDropHelper.stop());
        }
        else if (this.selection.dropTarget && !this.selection.dropTarget.locked) {
            if (this.selection.dropTarget instanceof xrReport_1.ReportSurface)
                return;
            var position = this._getAbsoluteSurfacePosition(ui);
            this.selection.dropTarget.underCursor().x = position.left - this.selection.dropTarget['absolutePosition'].x();
            this.selection.dropTarget.underCursor().y = position.top - this.selection.dropTarget['absolutePosition'].y();
            if (this.surface().isFit && this.surface().isFit(this.selection.dropTarget) || this.selection.dropTarget.underCursor().isOver) {
                if (draggable.data && draggable.data.specifics === 'stylemodel') {
                    this.selection.dropTarget.getControlModel()['styleName'] && this.selection.dropTarget.getControlModel()['styleName'](draggable.data.displayName);
                }
                else if (draggable.data && draggable.data.specifics === 'formattingrule') {
                    this.selection.dropTarget.getControlModel()['formattingRuleLinks'] && this.selection.dropTarget.getControlModel()['formattingRuleLinks'].push(formattingrules_1.FormattingRuleLink.createNew(draggable.data.data));
                }
            }
        }
    };
    return ReportExplorerDragDropHandler;
}(_objectExplorerDragDropHandler_1.ObjectExplorerDragDropHandler));
exports.ReportExplorerDragDropHandler = ReportExplorerDragDropHandler;
