// devexpress-reporting/designer/controls/metadata/properties/editOptions.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\properties\editOptions.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.textEditOptions = exports.editOptions = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var editOptions_1 = require("../../properties/editOptions");
exports.editOptions = {
    propertyName: 'editOptions',
    modelName: 'EditOptions',
    displayName: 'Edit Options', localizationId: 'DevExpress.XtraReports.UI.XRLabel.EditOptions',
    editor: analytics_widgets_1.editorTemplates.getEditor('objecteditor'),
    from: function (model, serializer) { return new editOptions_1.EditOptions(model, serializer); },
    toJsonObject: function (value, serializer) { return serializer.serialize(value); }
};
exports.textEditOptions = analytics_internal_1.extend({}, exports.editOptions, {
    propertyName: 'textEditOptions',
    from: function (model, serializer) { return new editOptions_1.TextEditOptions(model, serializer); }
});
