// devexpress-reporting/designer/controls/metadata/properties/formattingrules.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\properties\formattingrules.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formattingRuleSerializationsInfo = exports.formatting = exports.conditionObj = exports.formattingSerializationsInfo = exports.defaultBooleanVisible = exports.formattingRuleLinkSerializationsInfo = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var metadata_1 = require("../../../../common/metadata");
var metadata_2 = require("./metadata");
exports.formattingRuleLinkSerializationsInfo = [
    { propertyName: 'value', modelName: '@Value', link: true }
];
exports.defaultBooleanVisible = {
    propertyName: 'visible', modelName: '@Visible', displayName: 'Visible', localizationId: 'DevExpress.XtraReports.UI.Formatting.Visible', defaultVal: 'Default', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: metadata_2.defaultBooleanValuesArray
};
var borderWidthSerializable = { propertyName: 'borderWidthSerializable', modelName: '@BorderWidthSerializable', displayName: 'Border Width', localizationId: 'DevExpress.XtraReports.UI.Formatting.BorderWidthSerializable', from: analytics_utils_1.floatFromModel, editor: analytics_widgets_1.editorTemplates.getEditor('numeric') }, sides = analytics_internal_1.extend({ displayName: 'Borders', editor: analytics_widgets_1.editorTemplates.getEditor('borders'), localizationId: 'DevExpress.XtraReports.UI.XRControl.Borders' }, metadata_1.previewSides);
exports.formattingSerializationsInfo = [metadata_2.backColor, sides, metadata_2.borderColor, metadata_2.borderDashStyle, borderWidthSerializable,
    metadata_2.foreColor, metadata_2.font, metadata_2.textAlignment, exports.defaultBooleanVisible
].concat(metadata_2.paddingGroup);
exports.conditionObj = { propertyName: 'conditionObj', displayName: 'Condition', localizationId: 'DevExpress.XtraReports.UI.FormattingRule.Condition', editor: analytics_widgets_1.editorTemplates.getEditor('expressionEditor') };
exports.formatting = { propertyName: 'formatting', modelName: 'Formatting', displayName: 'Formatting', localizationId: 'DevExpress.XtraReports.UI.FormattingRule.Formatting', info: exports.formattingSerializationsInfo, editor: analytics_widgets_1.editorTemplates.getEditor('objecteditor') };
exports.formattingRuleSerializationsInfo = [
    { propertyName: 'name', modelName: '@Name', displayName: 'Name', localizationId: 'DevExpress.XtraReports.UI.FormattingRule.Name', editor: analytics_widgets_1.editorTemplates.getEditor('text'), validationRules: analytics_internal_1.nameValidationRules },
    { propertyName: 'condition', modelName: '@Condition', displayName: 'Condition', localizationId: 'DevExpress.XtraReports.UI.FormattingRule.Condition', defaultVal: '' },
    exports.conditionObj,
    metadata_2.dataSource, metadata_2.dataMember,
    exports.formatting
];
