// @devexpress/analytics-core/query-builder/dataSource/utils.js
/**
* DevExpress Analytics (query-builder\dataSource\utils.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FederationQueryType = exports.JsonSourceType = exports.SqlQueryType = void 0;
exports.SqlQueryType = {
    customSqlQuery: 'CustomSqlQuery',
    tableQuery: 'SelectQuery',
    storedProcQuery: 'StoredProcQuery'
};
exports.JsonSourceType = {
    fileJsonSource: 'FileJsonSource',
    customJsonSource: 'CustomJsonSource',
    uriJsonSource: 'UriJsonSource'
};
var FederationQueryType;
(function (FederationQueryType) {
    FederationQueryType[FederationQueryType["SelectNode"] = 0] = "SelectNode";
    FederationQueryType[FederationQueryType["UnionNode"] = 1] = "UnionNode";
    FederationQueryType[FederationQueryType["SourceNode"] = 2] = "SourceNode";
    FederationQueryType[FederationQueryType["TransformationNode"] = 3] = "TransformationNode";
})(FederationQueryType = exports.FederationQueryType || (exports.FederationQueryType = {}));
