// devexpress-reporting/chart/components/series/_template.js
/**
* DevExpress HTML/JS Reporting (chart\components\series\_template.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.seriesTemplate = exports.seriesTemplateSerializationsInfo = exports.valueDataMembersSerializable = exports.argumentDataMember = exports.labelsVisibility = exports.valueScaleType = exports.argumentScaleType = exports._argumentScaleTypeValidatorOptions = exports.legendTextPattern = exports.seriesPointsSortingKey = exports.seriesPointsSorting = exports.SeriesTemplateViewModel = exports.ScaleTypeMap = exports.ScaleType = void 0;
var analytics_elements_1 = require("@devexpress/analytics-core/analytics-elements");
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var ko = require("knockout");
var _commonValue_1 = require("../../internal/data/_commonValue");
var _axis_1 = require("../../internal/meta/_axis");
var _chart_1 = require("../../internal/meta/_chart");
var _common_1 = require("../../internal/meta/_common");
var _editorTemplates_1 = require("../../internal/_editorTemplates");
var _utils_1 = require("../../internal/_utils");
var _label_1 = require("./_label");
var _templateOptions_1 = require("./_templateOptions");
var _view_1 = require("./_view");
var _isBubble = function (viewType) {
    return viewType === 'BubbleSeriesView';
};
var _isRange = function (viewType) {
    return viewType === 'PolarRangeAreaSeriesView' ||
        viewType === 'RadarRangeAreaSeriesView' ||
        viewType === 'RangeArea3DSeriesView' ||
        viewType === 'RangeAreaSeriesView' ||
        viewType === 'OverlappedGanttSeriesView' ||
        viewType === 'SideBySideGanttSeriesView' ||
        viewType === 'OverlappedRangeBarSeriesView' ||
        viewType === 'SideBySideRangeBarSeriesView';
};
var _isStock = function (viewType) {
    return viewType === 'StockSeriesView' ||
        viewType === 'CandleStickSeriesView';
};
var ScaleType;
(function (ScaleType) {
    ScaleType[ScaleType["Qualitative"] = 0] = "Qualitative";
    ScaleType[ScaleType["Numerical"] = 1] = "Numerical";
    ScaleType[ScaleType["DateTime"] = 2] = "DateTime";
    ScaleType[ScaleType["Auto"] = 3] = "Auto";
})(ScaleType = exports.ScaleType || (exports.ScaleType = {}));
exports.ScaleTypeMap = {
    'Qualitative': ScaleType.Qualitative,
    'Numerical': ScaleType.Numerical,
    'DateTime': ScaleType.DateTime,
    'Auto': ScaleType.Auto
};
function _fixSeriesTemplateModel(model) {
    delete model['@SeriesDataMember'];
    return model;
}
var SeriesTemplateViewModel = (function (_super) {
    __extends(SeriesTemplateViewModel, _super);
    function SeriesTemplateViewModel(model, serializer, info) {
        var _this = _super.call(this, _fixSeriesTemplateModel(model), serializer, info) || this;
        _this._actualArgumentScaleType = ko.observable(ScaleType.Numerical);
        if (_this.valueDataMembers) {
            ko.unwrap(_this.valueDataMembers).dispose();
        }
        _this.viewType = ko.observable(_this.view().typeName);
        _this._adjustArgumentScaleType();
        var newInfo = analytics_internal_1.extend(true, [], info);
        var seriesPointsSortingKeyMeta = newInfo.filter(function (x) { return x.propertyName === exports.seriesPointsSortingKey.propertyName; })[0];
        Object.defineProperty(seriesPointsSortingKeyMeta, 'valuesArray', {
            get: function () { return _this._getCurrentSeriesPointsSortingKeys(_this.viewType()); },
            configurable: true
        });
        _this.viewType.subscribe(function (newType) {
            var newSerializer = serializer || new analytics_utils_1.ModelSerializer();
            _this.view(_view_1.SeriesViewViewModel.from({ '@TypeNameSerializable': newType }, newSerializer)());
            _this._adjustArgumentScaleType();
            _this._updateSeriesPointsSortingKey(seriesPointsSortingKeyMeta);
        });
        _this.getInfo = function () { return newInfo; };
        _this._disposables.push(_this.view.subscribe(function (newView) {
            if (_this.viewType() !== newView.typeName) {
                _this.viewType(newView.typeName);
            }
            _this.updateByView(newView);
        }));
        _this.valueDataMembers = ko.observable(new (_templateOptions_1.viewTypesDataMembers[_this.view().typeName] || _commonValue_1.CommonValueDataMembers)(model['@ValueDataMembersSerializable'] || '', _this.valueScaleType));
        _this._disposables.push(_this['indicators'] = ko.computed(function () {
            return _this.view() && _this.view()['indicators'] && _this.view()['indicators']();
        }));
        _this._disposables.push(_this['titles'] = ko.computed(function () {
            return _this.view() && _this.view()['titles'] && _this.view()['titles']();
        }));
        _this._disposables.push(_this._actualArgumentScaleType.subscribe(function (newVal) {
            switch (newVal) {
                case ScaleType.DateTime:
                    _this.qualitativeSummaryOptions.resetAllProperties();
                    _this.numericSummaryOptions.resetAllProperties();
                    break;
                case ScaleType.Numerical:
                    _this.qualitativeSummaryOptions.resetAllProperties();
                    _this.dateTimeSummaryOptions.resetAllProperties();
                    break;
                case ScaleType.Qualitative:
                    _this.numericSummaryOptions.resetAllProperties();
                    _this.dateTimeSummaryOptions.resetAllProperties();
                    break;
            }
        }));
        _this.updateByView(_this.view());
        [_this.qualitativeSummaryOptions, _this.numericSummaryOptions, _this.dateTimeSummaryOptions].forEach(function (options) {
            options['getPath'] = function (propertyName) {
                return _this['getPath'] && _this['getPath']('summaryFunction') || '';
            };
        });
        _this.viewBindable = {
            model: _this.view,
            type: _this.viewType
        };
        _this.filterString = new analytics_widgets_1.FilterStringOptions(_this._filterString, ko.pureComputed(function () { return _this['getPath'] && _this['getPath']('argumentDataMember') || ''; }));
        return _this;
    }
    SeriesTemplateViewModel.from = function (model, serializer) {
        return new SeriesTemplateViewModel(model || {}, serializer, exports.seriesTemplateSerializationsInfo);
    };
    SeriesTemplateViewModel.toJson = function (value, serializer, refs) {
        return serializer.serialize(value, exports.seriesTemplateSerializationsInfo, refs);
    };
    SeriesTemplateViewModel.prototype.updateByView = function (view) {
        if (this.label && this.label.typeNameSerializable) {
            this.label.typeNameSerializable(_templateOptions_1.mapTypes[view.typeName]);
        }
        this.valueDataMembers.peek().dispose();
        this.valueDataMembers(new (_templateOptions_1.viewTypesDataMembers[view.typeName] || _commonValue_1.CommonValueDataMembers)(this.valueDataMembers.peek().toString(), this.valueScaleType));
    };
    SeriesTemplateViewModel.prototype.preInitProperties = function (model) {
        delete model.DataFilters;
    };
    SeriesTemplateViewModel.prototype._isOnlyNumericArgumentScaleTypeSupported = function () {
        return _templateOptions_1.onlyNumericArgumentSupportedSeriesViewTypes.indexOf(this.viewType()) > -1;
    };
    SeriesTemplateViewModel.prototype._getCurrentSeriesPointsSortingKeys = function (viewType) {
        if (_isBubble(viewType))
            return _chart_1.bubbleSeriesPointsSortingKeys;
        if (_isRange(viewType))
            return _chart_1.rangeSeriesPointsSortingKeys;
        if (_isStock(viewType))
            return _chart_1.stockSeriesPointsSortingKeys;
        return _chart_1.commonSeriesPointsSortingKeys;
    };
    SeriesTemplateViewModel.prototype._adjustArgumentScaleType = function () {
        if (this._isOnlyNumericArgumentScaleTypeSupported())
            this.argumentScaleType('Numerical');
    };
    SeriesTemplateViewModel.prototype._updateSeriesPointsSortingKey = function (seriesPointsSortingKeyMeta) {
        var value = this[seriesPointsSortingKeyMeta.propertyName];
        if (!!value() && !seriesPointsSortingKeyMeta.valuesArray.some(function (x) { return x.value === value(); })) {
            value(seriesPointsSortingKeyMeta.valuesArray[0].value);
        }
    };
    SeriesTemplateViewModel.prototype._isDataMemberPropertyDisabled = function (name) {
        if (SeriesTemplateViewModel.dataMemberProperies.concat(['valueScaleType', 'argumentScaleType']).some(function (x) { return x === name; })) {
            return !ko.unwrap(this.dataSource);
        }
        return false;
    };
    SeriesTemplateViewModel.prototype._isPropertyDisabled = function (name) {
        if (name === 'argumentScaleType') {
            return this._isOnlyNumericArgumentScaleTypeSupported();
        }
        return false;
    };
    SeriesTemplateViewModel.prototype.getPath = function (propertyName) {
        var dataSourceInfo = ko.unwrap(this.dataSource);
        if (dataSourceInfo) {
            var helper = dataSourceInfo.dsHelperProvider && dataSourceInfo.dsHelperProvider();
            return helper && helper.getDataSourcePath(dataSourceInfo);
        }
    };
    SeriesTemplateViewModel.prototype.isPropertyVisible = function (propertyName) {
        switch (propertyName) {
            case 'qualitativeSummaryOptions':
                return this._actualArgumentScaleType() === ScaleType.Qualitative;
            case 'numericSummaryOptions':
                return this._actualArgumentScaleType() === ScaleType.Numerical;
            case 'dateTimeSummaryOptions':
                return this._actualArgumentScaleType() === ScaleType.DateTime;
            default:
                return true;
        }
    };
    SeriesTemplateViewModel.dataMemberProperies = ['argumentDataMember', 'valueDataMembers'];
    return SeriesTemplateViewModel;
}(analytics_elements_1.SerializableModel));
exports.SeriesTemplateViewModel = SeriesTemplateViewModel;
exports.seriesPointsSorting = {
    propertyName: 'seriesPointsSorting', modelName: '@SeriesPointsSorting', displayName: 'Series Points Sorting', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: [{ value: 'None', displayValue: 'None', localizationId: 'DevExpress.XtraReports.UI.MultiColumnMode.None' }, { value: 'Ascending', displayValue: 'Ascending', localizationId: 'DevExpress.XtraReports.UI.XRColumnSortOrder.Ascending' }, { value: 'Descending', displayValue: 'Descending', localizationId: 'DevExpress.XtraReports.UI.XRColumnSortOrder.Descending' }],
    localizationId: 'DevExpress.XtraCharts.SeriesBase.SeriesPointsSorting'
}, exports.seriesPointsSortingKey = {
    propertyName: 'seriesPointsSortingKey', modelName: '@SeriesPointsSortingKey', displayName: 'Series Points Sorting Key', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: _chart_1.commonSeriesPointsSortingKeys, localizationId: 'DevExpress.XtraCharts.SeriesBase.SeriesPointsSortingKey'
}, exports.legendTextPattern = { propertyName: 'legendTextPattern', modelName: '@LegendTextPattern', displayName: 'Legend Text Pattern', editor: analytics_widgets_1.editorTemplates.getEditor('text'), localizationId: 'DevExpress.XtraCharts.SeriesBase.LegendTextPattern' }, exports._argumentScaleTypeValidatorOptions = {
    _seriesViewModel: null,
    onInitialized: function (e) { exports._argumentScaleTypeValidatorOptions._seriesViewModel = e.model._model; },
    validationRules: [{
            type: 'custom',
            reevaluate: true,
            validationCallback: function (params) {
                var model = exports._argumentScaleTypeValidatorOptions._seriesViewModel.peek();
                if ('points' in model && model instanceof SeriesTemplateViewModel) {
                    model = model;
                    var unconvertiblePoint = _utils_1._getUnconvertiblePoint('argumentSerializable', model.argumentScaleType.peek(), params.value, model.points.peek());
                    var stringFormat = analytics_utils_1.getLocalization("The type of the '{0}' point isn't compatible with the {1} scale.", 'ChartStringId.MsgIncompatiblePointType');
                    var argumentValue = unconvertiblePoint && unconvertiblePoint.argumentSerializable();
                    if (argumentValue && (argumentValue instanceof Date))
                        argumentValue = analytics_internal_1.formatDate(argumentValue);
                    params.rule.message = analytics_internal_1.formatUnicorn(stringFormat, (argumentValue !== null || argumentValue !== void 0) ? argumentValue : '', params.value);
                    return !unconvertiblePoint;
                }
                return true;
            }
        }]
}, exports.argumentScaleType = {
    propertyName: 'argumentScaleType', modelName: '@ArgumentScaleType', displayName: 'Argument Scale Type', defaultVal: 'Auto', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: _common_1.scaleTypeValues, localizationId: 'DevExpress.XtraCharts.SeriesBase.ArgumentScaleType', validatorOptions: exports._argumentScaleTypeValidatorOptions
}, exports.valueScaleType = {
    propertyName: 'valueScaleType', modelName: '@ValueScaleType', displayName: 'Value Scale Type', defaultVal: 'Numerical', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: [
        { value: 'Numerical', displayValue: 'Numerical', localizationId: 'DevExpress.XtraCharts.ScaleType.Numerical' },
        { value: 'DateTime', displayValue: 'DateTime', localizationId: 'DevExpress.XtraTreeList.Data.UnboundColumnType.DateTime' }
    ],
    localizationId: 'DevExpress.XtraCharts.SeriesBase.ValueScaleType'
}, exports.labelsVisibility = { propertyName: 'labelsVisibility', modelName: '@LabelsVisibility', displayName: 'Labels Visibility', defaultVal: 'Default', editor: analytics_widgets_1.editorTemplates.getEditor('combobox'), valuesArray: _common_1.defaultBooleanValues, localizationId: 'DevExpress.XtraCharts.SeriesBase.LabelsVisibility' }, exports.argumentDataMember = { propertyName: 'argumentDataMember', modelName: '@ArgumentDataMember', displayName: 'Argument Data Member', defaultVal: '', editor: _editorTemplates_1.editorTemplates.getEditor('valueDataMember'), localizationId: 'DevExpress.XtraCharts.SeriesBase.ArgumentDataMember' }, exports.valueDataMembersSerializable = { propertyName: 'valueDataMembers', modelName: '@ValueDataMembersSerializable', displayName: 'Value Data Members', defaultVal: '', editor: analytics_widgets_1.editorTemplates.getEditor('objecteditor'), from: _commonValue_1.CommonValueDataMembers.from, toJsonObject: _commonValue_1.CommonValueDataMembers.toJson, localizationId: 'DevExpress.XtraCharts.SeriesBase.ValueDataMembers' };
exports.seriesTemplateSerializationsInfo = [
    _chart_1.viewBindableSerializationInfo,
    { propertyName: 'titles', displayName: 'Titles', localizationId: 'DevExpress.XtraReports.UI.XRChart.Titles' },
    { propertyName: 'indicators', displayName: 'Indicators', localizationId: 'DevExpress.XtraReports.UI.XRChart.Indicators' },
    _view_1.view, exports.argumentDataMember, exports.valueDataMembersSerializable, _chart_1.colorDataMember, exports.argumentScaleType, exports.valueScaleType, _common_1.filterString,
    _common_1.filterStringEditable, exports.seriesPointsSorting, exports.seriesPointsSortingKey, _common_1.showInLegend, _axis_1.legendName, exports.legendTextPattern, exports.labelsVisibility,
    _chart_1.qualitativeSummaryOptions, _chart_1.numericSummaryOptions, _chart_1.dateTimeSumaryOptions,
    _label_1.seriesLabel, _chart_1.topNOptions, _common_1.visible, _editorTemplates_1.chartDataSource
];
exports.seriesTemplate = { propertyName: 'seriesTemplate', modelName: 'SeriesTemplate', displayName: 'Series Template', localizationId: 'DevExpress.XtraReports.UI.XRChart.SeriesTemplate', info: exports.seriesTemplateSerializationsInfo, from: SeriesTemplateViewModel.from, toJsonObject: SeriesTemplateViewModel.toJson, editor: analytics_widgets_1.editorTemplates.getEditor('objecteditor') };
