// devexpress-reporting/designer/utils/settings.js
/**
* DevExpress HTML/JS Reporting (designer\utils\settings.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultCrossTabControl = exports.formatStringEditorCustomSet = exports.HandlerUri = exports.DataBindingMode = exports.smartTagFactory = exports.controlsFactory = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
exports.controlsFactory = analytics_internal_1.createGlobalModuleVariableFunc(null);
exports.smartTagFactory = analytics_internal_1.createGlobalModuleVariableFunc({});
exports.DataBindingMode = analytics_internal_1.createGlobalModuleVariableFunc('Expressions');
exports.HandlerUri = analytics_internal_1.createGlobalModuleVariableFunc('DXXRD.axd');
exports.formatStringEditorCustomSet = analytics_internal_1.createGlobalModuleVariableFunc({});
exports.DefaultCrossTabControl = analytics_internal_1.createGlobalModuleVariableFunc('XRCrossTab');
