// devexpress-reporting/designer/controls/metadata/xrPicturebox.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\xrPicturebox.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.popularPropertiesPicture = exports.pictureBoxSerializationsInfo = exports.imageEditOptions = exports.imageSource = exports.imageUrl = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var ko = require("knockout");
var imageSource_1 = require("../../../common/imageSource");
var editingField_1 = require("../../../viewer/editing/editingField");
var dataBinding_1 = require("../../dataObjects/metadata/dataBinding");
var editOptions_1 = require("../properties/editOptions");
var _metaUtils_1 = require("../utils/_metaUtils");
var anchoring_1 = require("./properties/anchoring");
var editOptions_2 = require("./properties/editOptions");
var metadata_1 = require("./properties/metadata");
var metadataGroups_1 = require("./properties/metadataGroups");
var scriptMetadata_1 = require("./properties/scriptMetadata");
exports.imageUrl = { propertyName: 'imageUrl', modelName: '@ImageUrl', editor: analytics_widgets_1.editorTemplates.getEditor('text'), defaultVal: '', displayName: 'Image Url', localizationId: 'DevExpress.XtraReports.UI.XRPictureBox.ImageUrl' };
exports.imageSource = {
    propertyName: 'imageSource',
    modelName: '@ImageSource',
    editor: { header: 'dxrd-image-loadfile', editorType: analytics_widgets_1.Editor },
    displayName: 'Image Source',
    localizationId: 'DevExpress.XtraReports.UI.XRPictureBox.ImageSource',
    from: function (val) { return ko.observable(imageSource_1.ImageSource.parse(val)); },
    toJsonObject: imageSource_1.ImageSource.toString,
    defaultVal: null
};
exports.imageEditOptions = analytics_internal_1.extend({}, editOptions_2.editOptions, {
    propertyName: 'imageEditOptions',
    from: function (model, serializer) { return new editOptions_1.ImageEditOptions(model, serializer); }
});
exports.pictureBoxSerializationsInfo = [
    exports.imageUrl, exports.imageSource, editingField_1.sizing, editingField_1.imageAlignment, metadata_1.keepTogether, anchoring_1.anchorVertical, anchoring_1.anchorHorizontal, scriptMetadata_1.controlScripts,
    dataBinding_1.dataBindings(['Bookmark', 'Image', 'ImageSource', 'ImageUrl', 'NavigateUrl', 'Tag']),
    exports.imageEditOptions
].concat(_metaUtils_1.createPopularBindingInfos({ propertyName: 'ImageUrl', localizationId: 'DevExpress.XtraReports.UI.XRPictureBox.ImageUrl' }), _metaUtils_1.createPopularBindingInfos({ propertyName: 'ImageSource', localizationId: 'DevExpress.XtraReports.UI.XRPictureBox.ImageSource' }), metadataGroups_1.sizeLocation, metadataGroups_1.commonControlProperties, metadataGroups_1.navigationGroup, metadataGroups_1.processGroup);
exports.popularPropertiesPicture = ['imageSource', 'popularDataBindingImageSource', 'imageUrl', 'popularDataBindingImageUrl', 'sizing', 'imageAlignment', 'bookmark', 'bookmarkParent'];
