// devexpress-reporting/designer/controls/metadata/xrSubreport.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\xrSubreport.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.subreportSerializationsInfo = exports.generateOwnPages = exports.parameterBindings = exports.reportSource = exports.reportSourceUrl = void 0;
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var editorTemplates_1 = require("../../widgets/editorTemplates");
var subreportViewModel_1 = require("../subreportViewModel");
var xrSubreportParameterBinding_1 = require("../xrSubreportParameterBinding");
var formattingRulesLink_1 = require("./properties/formattingRulesLink");
var metadata_1 = require("./properties/metadata");
var metadataGroups_1 = require("./properties/metadataGroups");
var scriptMetadata_1 = require("./properties/scriptMetadata");
exports.reportSourceUrl = { propertyName: 'reportSourceUrl', modelName: '@ReportSourceUrl', defaultVal: '', editor: editorTemplates_1.designerEditorTemplates.getEditor('reportSourceUrl'), displayName: 'Report Source Url', localizationId: 'DevExpress.XtraReports.UI.XRSubreport.ReportSourceUrl' };
exports.reportSource = {
    propertyName: 'reportSource', modelName: 'ReportSource',
    from: subreportViewModel_1.SubreportViewModel.from,
    toJsonObject: subreportViewModel_1.SubreportViewModel.toJson
};
exports.parameterBindings = { propertyName: 'parameterBindings', modelName: 'ParameterBindings', displayName: 'Parameter Bindings', localizationId: 'DevExpress.XtraReports.UI.XRSubreport.ParameterBindings', array: true, editor: analytics_widgets_1.editorTemplates.getEditor('commonCollection'), addHandler: xrSubreportParameterBinding_1.ParameterBinding.createNew, template: '#dxrd-commonCollectionItem' };
exports.generateOwnPages = { propertyName: 'generateOwnPages', defaultVal: false, modelName: '@GenerateOwnPages', displayName: 'Generate Own Pages', localizationId: 'DevExpress.XtraReports.UI.XRSubreport.GenerateOwnPages', editor: analytics_widgets_1.editorTemplates.getEditor('bool'), from: analytics_utils_1.parseBool };
exports.subreportSerializationsInfo = [
    exports.reportSource, exports.reportSourceUrl, scriptMetadata_1.subreportScripts, exports.parameterBindings, metadata_1.bookmarkParent, metadata_1.canShrink, exports.generateOwnPages, formattingRulesLink_1.formattingRuleLinks
].concat(metadataGroups_1.sizeLocation, metadataGroups_1.baseControlProperties.filter(function (x) { return x.modelName != metadata_1.tag.modelName; }));
