// devexpress-reporting/designer/controls/metadata/xrTable.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\xrTable.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.popularPropertiesTable = exports.tableSerializationsInfo = exports.processHiddenCellMode = void 0;
var analytics_widgets_1 = require("@devexpress/analytics-core/analytics-widgets");
var dataBinding_1 = require("../../dataObjects/metadata/dataBinding");
var anchoring_1 = require("./properties/anchoring");
var metadata_1 = require("./properties/metadata");
var metadataGroups_1 = require("./properties/metadataGroups");
var scriptMetadata_1 = require("./properties/scriptMetadata");
exports.processHiddenCellMode = {
    propertyName: 'processHiddenCellMode',
    modelName: '@ProcessHiddenCellMode',
    displayName: 'Process Hidden Cell Mode',
    valuesArray: [
        { value: 'LeaveEmptySpace', displayValue: 'Leave Empty Space' },
        { value: 'ResizeCellsEqually', displayValue: 'Resize Cells Equally' },
        { value: 'ResizeCellsProportionally', displayValue: 'Resize Cells Proportionally' },
        { value: 'StretchPreviousCell', displayValue: 'Stretch Previous Cell' },
        { value: 'StretchNextCell', displayValue: 'Stretch Next Cell' },
        { value: 'DecreaseTableWidth', displayValue: 'Decrease Table Width' }
    ],
    defaultVal: 'LeaveEmptySpace',
    editor: analytics_widgets_1.editorTemplates.getEditor('combobox'),
    localizationId: 'DevExpress.XtraReports.UI.XRTable.ProcessHiddenCellMode'
};
exports.tableSerializationsInfo = [
    metadata_1.expressionableFont, metadata_1.foreColor, metadata_1.keepTogetherDefaultValueFalse, anchoring_1.anchorVertical, anchoring_1.anchorHorizontal, scriptMetadata_1.commonScripts,
    { propertyName: 'rows', modelName: 'Rows', array: true },
    dataBinding_1.dataBindings(['Bookmark', 'Tag']),
    metadata_1.rtl,
    metadata_1.textAlignment,
    exports.processHiddenCellMode,
].concat(metadataGroups_1.sizeLocation, metadataGroups_1.commonControlProperties, metadataGroups_1.bookmarkGroup).filter(function (x) { return x != metadata_1.accessibleDescription; });
exports.popularPropertiesTable = ['bookmark', 'bookmarkParent'];
