// @devexpress/analytics-core/query-builder/dataSource/federation/federatedQueries/unionQuery.js
/**
* DevExpress Analytics (query-builder\dataSource\federation\federatedQueries\unionQuery.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnionQuery = exports.UnionTypes = exports.unionQuerySerializationsInfo = void 0;
var ko = require("knockout");
var utils_1 = require("../../utils");
var serializer_1 = require("../../../../serializer/serializer");
var utils_2 = require("../../../../serializer/utils");
var federatedQueriesContainer_1 = require("../federatedQueriesContainer");
var _utils_1 = require("../../../../serializer/_utils");
exports.unionQuerySerializationsInfo = [
    { propertyName: 'queryType', modelName: '@QueryType' },
    { propertyName: 'unionType', modelName: '@UnionType' },
    { propertyName: 'alias', modelName: '@Alias' },
    { propertyName: 'queries', modelName: 'UnionElements', array: true }
];
var UnionTypes;
(function (UnionTypes) {
    UnionTypes[UnionTypes["Union"] = 0] = "Union";
    UnionTypes[UnionTypes["UnionAll"] = 1] = "UnionAll";
})(UnionTypes = exports.UnionTypes || (exports.UnionTypes = {}));
var UnionQuery = (function (_super) {
    __extends(UnionQuery, _super);
    function UnionQuery(model, dataSources, serializer) {
        var _this = _super.call(this, model, dataSources, serializer) || this;
        (serializer || new serializer_1.ModelSerializer()).deserialize(_this, _utils_1.extend(model, { '@ItemType': 'Query' }));
        _this._disposables.push(_this.queryType = ko.pureComputed(function () { return utils_1.FederationQueryType[utils_1.FederationQueryType.UnionNode]; }));
        _this.queries = utils_2.deserializeArray(model['UnionElements'], function (item) { return _this.createQuery(item); });
        return _this;
    }
    UnionQuery.prototype.getInfo = function () {
        return exports.unionQuerySerializationsInfo;
    };
    UnionQuery.prototype.generateName = function () {
        return this.queries()[0] && this.queries()[0].alias().split('_').pop();
    };
    return UnionQuery;
}(federatedQueriesContainer_1.FederatedQueriesContainer));
exports.UnionQuery = UnionQuery;
