// devexpress-reporting/viewer/parameters/multiValuesHelper.js
/**
* DevExpress HTML/JS Reporting (viewer\parameters\multiValuesHelper.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiValuesHelper = void 0;
var analytics_internal_1 = require("@devexpress/analytics-core/analytics-internal");
var ko = require("knockout");
var MultiValuesHelper = (function () {
    function MultiValuesHelper(value, items, selectAllValues) {
        var _this = this;
        this.items = items;
        this.selectedItems = ko.observableArray([]);
        this.value = value;
        this.dataSource = items;
        var allValues;
        this.maxDisplayedTags = Math.min(((items && items.length) || 1) - 1, MultiValuesHelper.maxDisplayedTags);
        this.isSelectedAll = ko.pureComputed({
            read: function () { return _this.value().length == items.length; },
            write: function (selectAll) {
                var newValue = selectAll ? (allValues || (allValues = items.map(function (x) { return x.value; }))) : [];
                if (!analytics_internal_1.compareArrays(_this.value.peek(), newValue))
                    _this.value(newValue);
            }
        });
        if (selectAllValues)
            this.isSelectedAll(true);
    }
    MultiValuesHelper.maxDisplayedTags = 3;
    return MultiValuesHelper;
}());
exports.MultiValuesHelper = MultiValuesHelper;
