// devexpress-reporting/designer/controls/properties/controlParameter.js
/**
* DevExpress HTML/JS Reporting (designer\controls\properties\controlParameter.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControlParameter = void 0;
var analytics_utils_1 = require("@devexpress/analytics-core/analytics-utils");
var ko = require("knockout");
var dataBinding_1 = require("../../dataObjects/dataBinding");
var controlParameter_1 = require("../metadata/properties/controlParameter");
var ControlParameter = (function (_super) {
    __extends(ControlParameter, _super);
    function ControlParameter(model, serializer, _dataSourceHelper, _dataBindingsProvider) {
        var _this = _super.call(this, model, serializer) || this;
        _this._dataSourceHelper = _dataSourceHelper;
        _this._dataBindingsProvider = _dataBindingsProvider;
        _this.visible = ko.observable(true);
        _this.dataMemberInfo = ko.observable();
        _this.fakeBinding = _this;
        return _this;
    }
    ControlParameter.createNew = function () {
        return new ControlParameter({}, null);
    };
    ControlParameter.prototype.getInfo = function () {
        return controlParameter_1.controlParameterInfos;
    };
    ControlParameter.prototype.isEmpty = function () {
        return false;
    };
    ControlParameter.prototype.setDataMemberInfo = function (dataMemberInfo) {
        this.dataMemberInfo(dataMemberInfo);
    };
    Object.defineProperty(ControlParameter.prototype, "dataType", {
        get: function () {
            var _a;
            return (_a = this.dataMemberInfo()) === null || _a === void 0 ? void 0 : _a.dataType;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ControlParameter.prototype, "specifics", {
        get: function () {
            var _a;
            if (this.parameter())
                return this.parameter().specifics;
            return (_a = this.dataMemberInfo()) === null || _a === void 0 ? void 0 : _a.specifics;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(ControlParameter.prototype, "name", {
        get: function () {
            return this.parameterName();
        },
        enumerable: false,
        configurable: true
    });
    ControlParameter.prototype.generateValue = function (undoEngine, dataSourceHelper, dataSources, dataBindingsProvider) {
        var value = _super.prototype.generateValue.call(this, undoEngine, dataSourceHelper, dataSources);
        this.initDataMemberInfo(dataSourceHelper, dataBindingsProvider);
        return value;
    };
    ControlParameter.prototype.initDataMemberInfo = function (dataSourceHelper, dataBindingsProvider) {
        var _this = this;
        dataSourceHelper = dataSourceHelper || this._dataSourceHelper();
        dataBindingsProvider = dataBindingsProvider || this._dataBindingsProvider();
        if (dataSourceHelper && dataBindingsProvider && this.dataSource() && !this.dataMemberInfo()) {
            var dataSourcePath = dataSourceHelper.getDataSourcePath(this.dataSource());
            var dataMemberParts = (this.dataMember() || '').split('.');
            var valueMember = dataMemberParts.pop();
            var dataMember = dataMemberParts.join('.');
            var request = new analytics_utils_1.PathRequest(dataSourcePath + (dataMember ? '.' + dataMember : ''));
            dataBindingsProvider.getItems(request).done(function (result) {
                _this.dataMemberInfo(result.filter(function (x) { return x.name === valueMember; })[0]);
            });
        }
    };
    return ControlParameter;
}(dataBinding_1.DataBindingBase));
exports.ControlParameter = ControlParameter;
