// devexpress-reporting/designer/internal/errorPanel/_types.js
/**
* DevExpress HTML/JS Reporting (designer\internal\errorPanel\_types.js)
* Version:  22.1.5
* Build date: Sep 5, 2022
* Copyright (c) 2012 - 2022 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorType = exports.ErrorSource = void 0;
var ErrorSource;
(function (ErrorSource) {
    ErrorSource[ErrorSource["ReportCreation"] = 0] = "ReportCreation";
    ErrorSource[ErrorSource["ReportLayout"] = 1] = "ReportLayout";
    ErrorSource[ErrorSource["ReportExport"] = 2] = "ReportExport";
    ErrorSource[ErrorSource["ReportScripts"] = 3] = "ReportScripts";
})(ErrorSource = exports.ErrorSource || (exports.ErrorSource = {}));
var ErrorType;
(function (ErrorType) {
    ErrorType[ErrorType["Error"] = 0] = "Error";
    ErrorType[ErrorType["Warning"] = 1] = "Warning";
    ErrorType[ErrorType["Information"] = 2] = "Information";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
